src/librustc_passes/ast_validation.rs

```diff
@@ -702,6 +703,13 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
             .span_bug(mac.span, "macro invocation missed in expansion; did you forget to override \
                                  the relevant `fold_*()` method in `PlaceholderExpander`?");
     }
+
+    fn visit_fn_header(&mut self, header: &'a FnHeader) {
+        if header.asyncness.node.is_async() && self.session.rust_2015() {
+            struct_span_err!(self.session, header.asyncness.span, E0670,
+                             "fns cannot be delcared `async` in 2015 edition").emit();
```

Comment:
To avoid the awkward spelling "fns", perhaps something like the following would sound read more naturally:
```rust
"`async fn` is not permitted in the 2015 edition"
```